vendor/wp-now/src/symlink-manager.ts

```diff
@@ -0,0 +1,220 @@
+import { PHP, UnmountFunction } from '@php-wasm/universal';
+import fs from 'node:fs/promises';
+import { FileChangeInfo } from 'node:fs/promises';
+import path from 'path';
+import { createNodeFsMountHandler } from '@php-wasm/node';
+
+type mountedTarget = {
+	preExisting: boolean;
+	referenceCount: number;
+	unmountFunction: UnmountFunction;
+}
+
+/**
+ * Manages symlinks for a site and ensures they are properly mounted.
+ */
+export class SymlinkManager {
+
+	private symlinks: Map<string, string> = new Map<string,string>(); // Store relative path of symlink and its vfs realpath
+	private mountedTargets: Map<string, mountedTarget> = new Map<string, mountedTarget>(); // Store the target of the symlink and its reference count
+	private php: PHP;
+	readonly projectPath: string;
+	private watcher: AsyncIterable<FileChangeInfo<string>>;
+	private watcherAbortController: AbortController;
+
+	/**
+	 * Create a new symlink manager.
+	 * @param php
+	 * @param projectPath
+	 *
+	 */
+	constructor(php: PHP, projectPath: string) {
+		this.symlinks = new Map();
+		this.php = php;
+		this.projectPath = projectPath;
+	}
+
+	/**
+	 * Scan the project path for symlinks and mount their targets in the PHP runtime.
+	 */
+	async scanAndCreateSymlinks() {
+		for await( const symlink of this.collectSymlinks(this.projectPath) ) {
+			const relativePath = path.relative(this.projectPath, symlink);
+			await this.addSymlink(relativePath);
+		}
+	}
+
+	/**
+	 * Watches the project path for changes and update symlinks accordingly.
+	 */
+	startWatching() {
+		this.watcherAbortController = new AbortController();
+		const { signal } = this.watcherAbortController;
+		this.watcher = fs.watch( this.projectPath, { recursive: true, signal });
+
+		try {
+			(async () => {
+				for await (const change of this.watcher) {
+					await this.handleFileChange(change);
+				}
+			})();
+		} catch (err) {
+			console.error(err);
+		} finally {
+			this.watcher = null;
+			this.watcherAbortController = null;
+		}
+	}
+
+	/**
+	 * Handle a file change event.
+	 * @param change
+	 */
+	async handleFileChange(change: FileChangeInfo<string>) {
+		const { filename } = change;
+		// Check if  the file exists
+		const fullPath = path.join(this.projectPath, filename);
+		const fileExistsOnHost = await this.fileExists(fullPath);
+
+		// If the file does not exist, it was deleted. Check if it was a symlink and remove it
+		if ( ! fileExistsOnHost && this.symlinks.has(filename) ) {
+				await this.deleteMountedTarget(filename);
+				this.symlinks.delete(filename);
```

Comment:
Thanks, you are right. It makes more sense to delete the actual symlink entry from the map in the `deleteMountTarget` method because it always has to happen anyway.

I would not rename it to `removeSymlink` though because the symlink is not what we are really removing, we are removing the mount to which the symlink pointed. The symlink map is just a way to internally keep track of symlinks and their link targets.